research/conditional/utils/model_utils.py

```diff
@@ -49,90 +48,96 @@
 from research.conditional.moe_layers.expert_choice import ExpertChoiceFF
 from research.conditional.moe_layers.token_choice import TokenChoiceFF
 from research.conditional.moe_layers.ff_timed import FeedForwardTimed
+from torch.nn.parallel import DistributedDataParallel as DDP
 
 
-def make_loss_function(loss_checkpoint_chungs: int):
+def make_loss_and_backprop_function(loss_checkpoint_chungs: int):
     if loss_checkpoint_chungs == 0:
-        return calculate_llm_loss
+        return calculate_llm_loss_and_backward_pass
     else:
-        return partial(chungized_llm_loss, n_chungs=loss_checkpoint_chungs)
+        return partial(
+            chungized_llm_loss_and_backward_pass, n_chungs=loss_checkpoint_chungs
+        )
 
 
-def chungized_llm_loss(
+def chungized_llm_loss_and_backward_pass(
     batch: LLMBatch,
     model: torch.nn.Module,
     mixed_precision: bool,
     vocab_size: int,
+    scaler: torch.cuda.amp.GradScaler,
     n_chungs: int,
+    backward_pass: bool,
```

Comment:
from what I've seen in the code, this is not needed. Remove this argument and replace it with `model.train()` and `model.eval()`

Reply:
Done here, probably needs some cleanup in code calling this @Simontwice